### DISCUSSION TITLE
@OneToMany에서 fetch join과 페이지네이션 API 함께 사용하면 어떤 일이 발생할까요?

### DISCUSSION THREAD

**Discussion by le2sky** [2024-woowacourse-study/level-interview#201] [Spring Data]
뭘까용

**Comment by le2sky** [comment]
batch size

**Comment by nayonsoso** [comment]
### 들어가기 앞서
- fetch join 은 한번의 쿼리로 **연관된 엔티티를 한꺼번에 로드**한다.
- JPA에서 제공하는 페이징은 setFirstResult()와 setMaxResults() 메서드를 사용해서 **데이터베이스의 특정 부분만 로드**할 수 있게 해준다.
- 예를 들어, Pageable(1,5) 객체는 5개씩 나누어진 곳에서 1페이지에 해당하는 것을 가져오겠다는 뜻이다.

### xxToOne 관계에서의 Fetch Join와 페이징
- 테이블을 조인해도 데이터의 수가 변하지 않기 때문에 문제 없다.

### xxToMany 관계에서의 Fetch Join와 페이징
- 이 관계에서는 테이블을 조인하면 데이터의 수가 변한다. (자세한 원리는 fetch join 을 더 찾아보길 추천함) 
- 따라서 스프링에서는 전체 데이터를 메모리에 로드한 후 중복을 없애고 offset 과 limit 을 적용해서 페이징하기를 ‘시도’하지만, 결과적으로 페이징이 되지 않는다. (https://junhyunny.github.io/spring-boot/jpa/jpa-fetch-join-paging-problem/#21-기본-페이징-처리-및-default_batch_fetch_size-설정-사용)

### Fetch Join와 페이징을 같이 사용하기 위한 방법
- 앞서 언급했듯, toOne 관계에서 사용하면 된다.
- 또는 배치 사이즈를 설정함으로써 해결할 수 있다.


ref.
https://tecoble.techcourse.co.kr/post/2020-10-21-jpa-fetch-join-paging/
https://junhyunny.github.io/spring-boot/jpa/jpa-fetch-join-paging-problem/#21-기본-페이징-처리-및-default_batch_fetch_size-설정-사용
https://velog.io/@hyungzin0309/JPA-일-대-다-Fetch-Join-Pagination
https://wangtak.tistory.com/30

**Comment by gyungchan-jo** [comment]
`@OneToMany`에서 fetch join과 페이지네이션 API 함께 사용하면 fetch join이 적용되지 않고 DB를 Full Scan하여 페이징을 적용하는 상황이 발생하여 OutOfMemoryError가 발생할 수 있습니다.

이유는 만약 부모 엔티티가 3개 있고, 각각의 부모 엔티티는 연관된 자식 엔티티가 7개 존재한다고 가정해봅시다. 1:N 관계를 Join하면 총 21(3 * 7)개의 DB Row가 조회됩니다. 데이터의 수가 변경되기 때문에 단순하게 LIMIT 구문을 사용하는 쿼리로 페이지네이션을 적용하기 어렵습니다. 따라서 조회한 결과를 모두 메모리로 가져와서 JPA가 페이지네이션 계산을 진행합니다.

그래서 페이징할 때 사용하던 기존의 SQL LIMIT 구문이 등장하지 않고, 쿼리 결과를 전부 메모리에 적재한 뒤 어플리케이션 단에서 Pagination 작업을 수행한다는 경고 로그가 발생합니다.

이를 해결하기 위해 배치 사이즈를 설정할 수 있습니다.

**Comment by le2sky** [comment]
https://tecoble.techcourse.co.kr/post/2021-07-26-jpa-pageable/
